aea/skills/base.py

```diff
@@ -912,3 +765,356 @@ def _print_warning_message_for_non_declared_skill_components(
+
+_ComponentsHelperIndex = Dict[_SKILL_COMPONENT_TYPES, Dict[str, SkillComponent]]
+"""
+Helper index to store component instances.
+"""
+
+
+class _SkillComponentLoadingItem:  # pylint: disable=too-few-public-methods
+    """Class to represent a triple (component name, component configuration, component class)."""
+
+    def __init__(
+        self,
+        name: str,
+        config: SkillComponentConfiguration,
+        class_: Type[SkillComponent],
+        type_: _SKILL_COMPONENT_TYPES,
+    ):
+        """Initialize the item."""
+        self.name = name
+        self.config = config
+        self.class_ = class_
+        self.type_ = type_
+
+
+class _SkillComponentLoader:
+    """This class implements the loading policy for skill components."""
+
+    def __init__(
+        self, configuration: SkillConfig, skill_context: SkillContext, **kwargs: Any
+    ):
+        """Initialize the helper class."""
+        enforce(
+            configuration.directory is not None,
+            "Configuration not associated to directory.",
+        )
+        self.configuration = configuration
+        self.skill_directory = cast(Path, configuration.directory)
+        self.skill_context = skill_context
+        self.kwargs = kwargs
+
+        self._all_component_names: Set[str] = set()
+
+        self.skill = Skill(self.configuration, self.skill_context, **self.kwargs)
+        self.skill_dotted_path = f"packages.{self.configuration.public_id.author}.skills.{self.configuration.public_id.name}"
+
+    def load_skill(self) -> Skill:
+        """Load the skill."""
+        load_aea_package(self.configuration)
+        python_modules: Set[Path] = self._get_python_modules()
+        declared_component_classes: Dict[
+            _SKILL_COMPONENT_TYPES, Dict[str, SkillComponentConfiguration]
+        ] = self._get_declared_skill_component_configurations()
+        self._all_component_names = {
+            config.class_name
+            for _, types_ in declared_component_classes.items()
+            for _, config in types_.items()
+        }
+        component_classes_by_path: Dict[
+            Path, Set[Tuple[str, Type[SkillComponent]]]
+        ] = self._load_component_classes(python_modules)
+        component_loading_items = self._match_class_and_configurations(
+            component_classes_by_path, declared_component_classes
+        )
+        components = self._get_component_instances(component_loading_items)
+        self._update_skill(components)
+        return self.skill
+
+    def _update_skill(self, components: _ComponentsHelperIndex) -> None:
+        self.skill.handlers.update(
+            cast(Dict[str, Handler], components.get(Handler, {}))
+        )
+        self.skill.behaviours.update(
+            cast(Dict[str, Behaviour], components.get(Behaviour, {}))
+        )
+        self.skill.models.update(cast(Dict[str, Model], components.get(Model, {})))
+        self.skill._set_models_on_context()  # pylint: disable=protected-access
```

Comment:
Previously we only did the during skill init. Why is it now needed here? Conceptually it makes sense but I don't understand why it worked before.